src/providers/wallet.tsx

```diff
@@ -0,0 +1,215 @@
+        } catch (error) {
+          console.error(error);
+          setError(error as Error);
+        }
+      }
+
+      return false;
+    },
+    [getWalletBySource]
+  );
+
+  const disConnectWallet = useCallback(() => {
+    setSigner(null);
+    setAccounts([]);
+    setWalletToUse(null);
+  }, []);
+
+  useEffect(() => {
+    const sub$$ = from(accountsObs.subject.asObservable()).subscribe(setAccountsObsData);
+
+    return () => sub$$.unsubscribe();
+  }, []);
+
+  useEffect(() => {
+    const readOnlyAddress = new URL(window.location.href).searchParams.get('address');
+    const readOnly =
+      readOnlyAddress && isValidAddress(readOnlyAddress)
+        ? [
+            {
+              address: readOnlyAddress,
+              displayAddress: convertToSS58(readOnlyAddress, network.ss58Prefix),
+              meta: { name: 'Read-Only', source: SEARCH_PARAMS_SOURCE },
+            },
+          ]
+        : [];
+
+    setAccounts((prev) => {
+      const exist = prev.find(({ meta }) => meta.source === SEARCH_PARAMS_SOURCE);
+
+      return exist ? prev : [...prev, ...readOnly];
+    });
+  }, [network.ss58Prefix]);
+
+  useEffect(() => {
+    if (!walletToUse) {
+      return;
+    }
+
+    setSigner(walletToUse.signer);
+
+    const apiGenesisHash = api.genesisHash.toHex();
+
+    const sub$$ = walletToUse.accounts.subscribe((accs) => {
+      const extension = accs
+        .filter((acc) => !acc.genesisHash || acc.genesisHash === apiGenesisHash)
+        .map((acc) => {
+          const { address, genesisHash, name, type } = acc;
+
+          return {
+            address,
+            displayAddress: convertToSS58(address, network.ss58Prefix),
+            type,
+            meta: {
+              genesisHash,
+              name,
+              source: walletToUse.extensionName,
+            },
+          };
+        });
+
+      const keys = Object.keys(accountsObsData);
+      const extensionAddresses = accs.map((item) => item.address);
+      const sources = keys.filter((key) => !extensionAddresses.includes(key));
+
+      const locals: Account[] = sources.map((address) => {
+        const found = accs.find((item) => item.address === address);
+
+        return {
+          address,
+          displayAddress: convertToSS58(address, network.ss58Prefix),
+          type: found?.type,
+          json: accountsObsData[address].json,
+          meta: {
+            genesisHash: found?.genesisHash,
+            name: found?.name,
+            source: LOCAL_SOURCE,
+          },
+        };
+      });
+
+      setAccounts((prev) => {
+        const readOnly = prev.find(({ meta }) => meta.source === SEARCH_PARAMS_SOURCE);
+
+        return readOnly ? [...extension, ...locals, readOnly] : [...extension, ...locals];
+      });
+    });
+
+    return () => sub$$();
+  }, [walletToUse, network.ss58Prefix, api, accountsObsData]);
+
+  useEffect(() => {
+    const injecteds = window.injectedWeb3;
+
+    setSupportedWallets([
```

Comment:
The line from 154 to 186 seems to be an configuration， isolate it?

Reply:
isolated